### PR TITLE
DEV: better highlighting of mentions

### PR DESCRIPTION
**Pull request by jjaffeux**
This commit improves the hilight-ing of mentions in posts and chat messages.

- `@here` and `@all` will generate a `<a class="mention --wide">`
- bots will generate a `<a class="mention --bot">`
- current user will generate a `<a class="mention --current">`

To achieve this change the following value transformer has been added: "mentions-class". It will be run in posts and chat messages after the mention is rendered.

A bug were bots were not considered in mentioned users has also been fixed as part of this PR.

### CODE REVIEW

File: plugins/chat/lib/chat/parsed_mentions.rb [hunk 1]
```diff
@@ -86,7 +86,7 @@ def channel_members
     end
 
     def chat_users
-      User.distinct.joins(:user_option).real.where(user_options: { chat_enabled: true })
+      User.distinct.joins(:user_option).where(user_options: { chat_enabled: true })
```
Comment (jjaffeux) [inline plugins/chat/lib/chat/parsed_mentions.rb:89]: We were not returning mentioned bots because of this

File: plugins/chat/assets/javascripts/discourse/components/chat-message.gjs [hunk 2]
```diff
@@ -239,6 +238,38 @@ export default class ChatMessage extends Component {
...
+  decorateMentions(cooked) {
+    if (this.args.message.channel.allowChannelWideMentions) {
+      const wideMentions = [...cooked.querySelectorAll("span.mention")];
+      MENTION_KEYWORDS.forEach((keyword) => {
+        const mentions = wideMentions.filter((node) => {
+          return node.textContent.trim() === `@${keyword}`;
+        });
+
+        mentionsDecorators().forEach((decorator) => {
+          decorator(mentions, { username: keyword });
```
Comment (jjaffeux) [inline plugins/chat/assets/javascripts/discourse/components/chat-message.gjs:259]: not ideal as this is not a real user, but that seems robust enough

File: app/assets/javascripts/discourse/app/widgets/post-cooked.js [hunk 3]
```diff
@@ -378,38 +376,39 @@ export default class PostCooked {
...
     }
-  }
 
-  _rerenderUserStatusOnMentions() {
-    destroyUserStatusOnMentions();
-    this._post()?.mentioned_users?.forEach((user) =>
-      this._rerenderUserStatusOnMention(this.cookedDiv, user)
-    );
-  }
+    this._post()?.mentioned_users?.forEach((user) => {
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/widgets/post-cooked.js:384]: Most of the refactoring here is done to avoid iterating multiple times on the list of users, and also sharing the mentions nodes

File: app/assets/javascripts/discourse/app/widgets/post-cooked.js [hunk 4]
```diff
@@ -66,16 +67,15 @@ export default class PostCooked {
...
-    this._initUserStatusOnMentions();
-    this._decorateAndAdopt(cookedDiv);
+    this._insertQuoteControls(this.cookedDiv);
+    this._showLinkCounts(this.cookedDiv);
+    this._applySearchHighlight(this.cookedDiv);
+    this._decorateMentions();
+    this._decorateAndAdopt(this.cookedDiv);
 
-    return cookedDiv;
+    return this.cookedDiv;
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/widgets/post-cooked.js:78]: not part of this PR but that assignment seemed unecessary

File: app/assets/javascripts/discourse/app/widgets/post-cooked.js [hunk 5]
```diff
@@ -365,9 +365,7 @@ export default class PostCooked {
 
     if (
       (this.attrs.firstPost || this.attrs.embeddedPost) &&
-      this.ignoredUsers &&
-      this.ignoredUsers.length > 0 &&
-      this.ignoredUsers.includes(this.attrs.username)
+      this.ignoredUsers?.includes?.(this.attrs.username)
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/widgets/post-cooked.js:368]: also not part of the PR, but that seemed simpler

File: app/assets/javascripts/discourse/app/lib/transformer/registry.js [hunk 6]
```diff
@@ -7,6 +7,7 @@ export const VALUE_TRANSFORMERS = Object.freeze([
   // use only lowercase names
   "category-description-text",
   "category-display-name",
+  "mentions-class",
```
Comment (davidtaylorhq) [inline app/assets/javascripts/discourse/app/lib/transformer/registry.js:10]: Bikeshedding but... the transformer runs for each mention (singlular), and acts on an array of classes (plural). So I think it might be better as:

```
"mention-classes"
```
?

**Review by davidtaylorhq — APPROVED** [review]
